From 015b23b76ea6244b475704cb8e30475e31c162d5 Mon Sep 17 00:00:00 2001
From: "Chen, Yu" <yu.y.chen@intel.com>
Date: Sat, 22 Oct 2022 21:40:09 +0800
Subject: [PATCH] Vulkan: Add target device idx

Issue: Game "Guangyu" will crash when
       running in aic with non-zero rnode
Cause: Not specify the target device idx
       when initializing vulkan
Fix: Use property "ro.acg.rnode" to set
     default target device idx

Signed-off-by: Xiaohui Gu <xiaohui.gu@intel.com>
Signed-off-by: Marc Mao <marc.mao@intel.com>
Signed-off-by: Chen, Yu <yu.y.chen@intel.com>
Change-Id: Icbfacf144a79b482ea4599b6f46fcc56dcdc8ca0
---
 src/vulkan/runtime/vk_instance.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/src/vulkan/runtime/vk_instance.c b/src/vulkan/runtime/vk_instance.c
index 21a00674f54..185abe1689b 100644
--- a/src/vulkan/runtime/vk_instance.c
+++ b/src/vulkan/runtime/vk_instance.c
@@ -40,6 +40,11 @@
 #define VERSION_IS_1_0(version) \
    (VK_API_VERSION_MAJOR(version) == 1 && VK_API_VERSION_MINOR(version) == 0)
 
+#if defined(__ANDROID__) || defined(ANDROID)
+#include <cutils/properties.h>
+#define TARGET_DEVICE_IDX_PROP_NAME "ro.acg.rnode"
+#define DEFAULT_TARGET_DEVICE_IDX   "0"
+#endif
 VkResult
 vk_instance_init(struct vk_instance *instance,
                  const struct vk_instance_extension_table *supported_extensions,
@@ -370,6 +375,17 @@ enumerate_drm_physical_devices_locked(struct vk_instance *instance)
    VkResult result;
    for (uint32_t i = 0; i < (uint32_t)max_devices; i++) {
       struct vk_physical_device *pdevice;
+#if defined(__ANDROID__) || defined(ANDROID)
+      const char *path = devices[i]->nodes[DRM_NODE_RENDER];
+      char target_device_path[128];
+      char target_device_idx[PROPERTY_VALUE_MAX];
+      property_get(TARGET_DEVICE_IDX_PROP_NAME, target_device_idx, DEFAULT_TARGET_DEVICE_IDX);
+      snprintf(target_device_path, sizeof(target_device_path),"/dev/dri/renderD%d", atoi(target_device_idx)+128);
+      target_device_path[sizeof(target_device_path)-1] = '\0';
+      if(strncmp(path, target_device_path, strlen(target_device_path))){
+         continue;
+      }
+#endif
       result = instance->physical_devices.try_create_for_drm(instance, devices[i], &pdevice);
 
       /* Incompatible DRM device, skip. */
-- 
2.33.1

